// sequelize-rc/lib/test/models/users.js
'use strict';

exports.model = function(sequelize, DataTypes) {
    return sequelize.define('user', {
        id: {
            type: DataTypes.INTEGER,
            allowNull: false,
            autoIncrement: true,
            primaryKey: true
        },
        name: {
            type: DataTypes.STRING,
            allowNull: false
        },
        password: {
            type: DataTypes.STRING,
            allowNull: false
        },
        status: {
            type: DataTypes.STRING,
            allowNull: true
        },
        email: {
            type: DataTypes.STRING,
            allowNull: true
        },
        mobile: {
            type: DataTypes.STRING,
            allowNull: true
        },
        phone: {
            type: DataTypes.STRING,
            allowNull: true
        },
        public_key: {
            type: DataTypes.STRING,
            allowNull: true
        },
        twcn_auth: {
            type: DataTypes.TEXT,
            allowNull: true
        },
        enable: {
            defaultValue: 1,
            type: DataTypes.BOOLEAN,
            allowNull: false
        },
    }, {
        underscored: false,
        tableName: 'user'
    });
};
